packages/kmod-5.10-nvidia/nvidia-persistenced.service.in

```diff
@@ -0,0 +1,40 @@
+# NVIDIA Persistence Daemon Init Script -- Modified
+#
+# Copyright (c) 2013 NVIDIA Corporation
+#
+# Permission is hereby granted, free of charge, to any person obtaining a
+# copy of this software and associated documentation files (the "Software"),
+# to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense,
+# and/or sell copies of the Software, and to permit persons to whom the
+# Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+#
+# This is a sample systemd service file, designed to show how the NVIDIA
+# Persistence Daemon can be started.
+#
+
+[Unit]
+Description=NVIDIA Persistence Daemon
+Requires=load-kernel-modules.service
+After=load-kernel-modules.service
+
+[Service]
+Type=forking
```

Comment:
I ran some tests when I set the mode to simple, it looks like the unit immediately exits. 

```
○ nvidia-persistenced.service - NVIDIA Persistence Daemon
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/nvidia-persistenced.service; enabled; preset: enabled)
     Active: inactive (dead) since Fri 2024-09-06 17:24:31 UTC; 1min 50s ago
   Duration: 2.102s
   Main PID: 1424 (code=exited, status=0/SUCCESS)
        CPU: 2.814s

Sep 06 17:24:28 ip-172-31-92-66.ec2.internal nvidia-smi[1407]: |  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
Sep 06 17:24:28 ip-172-31-92-66.ec2.internal nvidia-smi[1407]: |        ID   ID                                                             Usage      |
Sep 06 17:24:28 ip-172-31-92-66.ec2.internal nvidia-smi[1407]: |=======================================================================================|
Sep 06 17:24:28 ip-172-31-92-66.ec2.internal nvidia-smi[1407]: |  No running processes found                                                           |
Sep 06 17:24:28 ip-172-31-92-66.ec2.internal nvidia-smi[1407]: +---------------------------------------------------------------------------------------+
Sep 06 17:24:29 ip-172-31-92-66.ec2.internal systemd[1]: Started NVIDIA Persistence Daemon.
Sep 06 17:24:29 ip-172-31-92-66.ec2.internal nvidia-persistenced[1426]: Started (1426)
Sep 06 17:24:31 ip-172-31-92-66.ec2.internal nvidia-persistenced[1426]: Shutdown (1426)
Sep 06 17:24:31 ip-172-31-92-66.ec2.internal systemd[1]: nvidia-persistenced.service: Deactivated successfully.
Sep 06 17:24:31 ip-172-31-92-66.ec2.internal systemd[1]: nvidia-persistenced.service: Consumed 2.814s CPU time.
bash-5.1# ps axu | grep persistenced
root        3274  0.0  0.0   3056  1508 pts/1    S+   17:55   0:00 grep persistenced
```

It seems like it only spawns the daemon when I run `nvidia-persistenced` from the command line.

```
bash-5.1# ./x86_64-bottlerocket-linux-gnu/sys-root/usr/libexec/nvidia/tesla/bin/nvidia-persistenced
bash-5.1# ps axu | grep persistenced
root        3279 48.7  0.0   4972  1660 ?        Ss   17:56   0:02 ./x86_64-bottlerocket-linux-gnu/sys-root/usr/libexec/nvidia/tesla/bin/nvidia-persistenced
root        3289  0.0  0.0   3056  1476 pts/1    S+   17:56   0:00 grep persistenced
```

I don't think simple is the correct unit type here. Maybe `notify`? Nvidia's provided template chooses `forking`.

Reply:
I read the `nvidia-persistenced` code, and the main parent process forks the daemon in a child process. This comment was helpful to understand the behavior:

https://github.com/NVIDIA/nvidia-persistenced/blob/39e71b2cb7542d14f5106043fb4c7a789428144b/nvidia-persistenced.c#L220

`forking` seems to be the only choice we have since `nvidia-persistenced` doesn't integrate with `sd_notify`.